specutils/manipulation/resample.py

```diff
@@ -0,0 +1,409 @@
+
+    >>> import numpy as np
+    >>> import astropy.units as u
+    >>> from specutils import Spectrum1D
+    >>> from specutils.manipulation import FluxConservingResample
+    >>> input_spectra = Spectrum1D(
+    ...     flux=np.array([1, 3, 7, 6, 20]) * u.mJy,
+    ...     spectral_axis=np.array([2, 4, 12, 16, 20]) * u.nm)
+    >>> resample_grid = np.array([1, 5, 9, 13, 14, 17, 21, 22, 23])
+    >>> fluxc_resample = FluxConservingResample()
+    >>> output_spectrum1D = fluxc_resample(input_spectra, resample_grid) # doctest: +IGNORE_OUTPUT
+
+    """
+
+    def __call__(self, orig_spectrum, fin_lamb):
+        """
+        Return the resulting `~specutils.Spectrum1D` of the resampling.
+        """
+        return self.resample1d(orig_spectrum, fin_lamb)
+
+    def _resample_matrix(self, orig_lamb, fin_lamb):
+        """
+        Create a re-sampling matrix to be used in re-sampling spectra in a way
+        that conserves flux. This is adapted from
+        https://ui.adsabs.harvard.edu/abs/2017arXiv170505165C/references,
+        eprint arXiv:1705.05165. This code was heavily influenced by Nick Earl's
+        resample rough draft.
+
+        Parameters
+        ----------
+        orig_lamb : ndarray
+            The original dispersion array.
+        fin_lamb : ndarray
+            The desired dispersion array.
+
+        Returns
+        -------
+        resample_mat : ndarray
+            An [[N_{fin_lamb}, M_{orig_lamb}]] matrix.
+        """
+        # Lower bin and upper bin edges
+        orig_edges = self._calc_bin_edges(orig_lamb)
+        fin_edges = self._calc_bin_edges(fin_lamb)
+
+        # I could get rid of these alias variables,
+        # but it does add readability
+        orig_low = orig_edges[:-1]
+        fin_low = fin_edges[:-1]
+        orig_upp = orig_edges[1:]
+        fin_upp = fin_edges[1:]
+
+        # Here's the real work in figuring out the bin overlaps
+        # i.e., contribution of each original bin to the resampled bin
+        l_inf = np.where(orig_low > fin_low[:, np.newaxis],
+                         orig_low, fin_low[:, np.newaxis])
+        l_sup = np.where(orig_upp < fin_upp[:, np.newaxis],
+                         orig_upp, fin_upp[:, np.newaxis])
+
+        resamp_mat = (l_sup - l_inf).clip(0)
+        resamp_mat *= (orig_upp - orig_low)
+
+        # set bins that don't overlap 100% with original bins
+        # to zero by checking edges, and applying generated mask
+        left_clip = np.where(fin_edges[:-1] - orig_edges[0] < 0, 0, 1)
+        right_clip = np.where(orig_edges[-1] - fin_edges[1:] < 0, 0, 1)
+        keep_overlapping_matrix = left_clip * right_clip
+
+        resamp_mat *= keep_overlapping_matrix[:, np.newaxis]
+
+        return resamp_mat
+
+    def resample1d(self, orig_spectrum, fin_lamb):
+        """
+        Create a re-sampling matrix to be used in re-sampling spectra in a way
+        that conserves flux. This is adapted from eprint arXiv:1705.05165. If
```

Comment:
I think there's no need to reference the paper twice.  It's better to just say it at the top of your class (as you have already), so that this docstring is about the actual method rather than the algorithm.